relayer_base/src/gmp_api/gmp_api_db_audit_decorator.rs

```diff
@@ -0,0 +1,628 @@
+use relayer_base::models::gmp_events::{GMPAudit, PgGMPEvents};
+use relayer_base::models::gmp_tasks::{GMPTaskAudit, PgGMPTasks};
+use relayer_base::config::Config;
+use sqlx::PgPool;
+use std::sync::Arc;
+
+async fn create(config: &Config, pg_pool: PgPool) -> anyhow::Result<impl GmpApiTrait> {
+    // Create the GMP API
+    let gmp_api = GmpApi::new(config, true)?;
+
+    // Create both models using the provided PgPool
+    let gmp_tasks = PgGMPTasks::new(pg_pool.clone());
+    let gmp_events = PgGMPEvents::new(pg_pool);
+
+    // Create a decorated GMP API with database auditing
+    let decorated_api = GmpApiDbAuditDecorator::new(gmp_api, gmp_tasks, gmp_events);
+
+    // Now you can use decorated_api as a regular GmpApiTrait implementation
+    // with automatic database auditing for tasks and events 
+    Ok(decorated_api)
+}
+```
+*/
+
+use crate::config::Config;
+use crate::error::GmpApiError;
+use crate::gmp_api::gmp_types::{
+    BroadcastRequest, CannotExecuteMessageReason, Event, PostEventResult, QueryRequest, Task,
+};
+use crate::gmp_api::{GmpApi, GmpApiTrait};
+use crate::models::gmp_events::{EventModel, GMPAudit, PgGMPEvents};
+use crate::models::gmp_tasks::{GMPTaskAudit, PgGMPTasks, TaskModel};
+use sqlx::{types::Json, PgPool};
+use std::sync::Arc;
+use tokio::spawn;
+use tracing::error;
+use xrpl_amplifier_types::msg::XRPLMessage;
+
+pub struct GmpApiDbAuditDecorator<T: GmpApiTrait, U: GMPTaskAudit, V: GMPAudit> {
+    gmp_api: T,
+    gmp_tasks: Arc<U>,
+    gmp_events: Arc<V>,
+}
+
+impl<T: GmpApiTrait, U: GMPTaskAudit, V: GMPAudit> GmpApiDbAuditDecorator<T, U, V> {
+    pub fn new(gmp_api: T, gmp_tasks: U, gmp_events: V) -> Self {
+        Self {
+            gmp_api,
+            gmp_tasks: Arc::new(gmp_tasks),
+            gmp_events: Arc::new(gmp_events),
+        }
+    }
+}
+
+/// Constructs a GmpApiDbAuditDecorator with GmpApi, PgGMPTasks, and PgGMPEvents
+///
+/// # Arguments
+/// 
+/// * `pg_pool` - A PostgreSQL connection pool
+/// * `config` - The configuration for the GmpApi
+/// * `connection_pooling` - Whether to enable connection pooling for the GmpApi
+/// 
+/// # Returns
+/// 
+/// A Result containing an Arc-wrapped GmpApiDbAuditDecorator or a GmpApiError
+pub fn construct_gmp_api(
+    pg_pool: PgPool,
+    config: &Config,
+    connection_pooling: bool,
+) -> Result<Arc<GmpApiDbAuditDecorator<GmpApi, PgGMPTasks, PgGMPEvents>>, GmpApiError> {
+    let gmp_tasks = PgGMPTasks::new(pg_pool.clone());
+    let gmp_events = PgGMPEvents::new(pg_pool);
+
+    let gmp_api_base = GmpApi::new(config, connection_pooling)?;
+    let gmp_api = Arc::new(GmpApiDbAuditDecorator::new(
+        gmp_api_base,
+        gmp_tasks,
+        gmp_events,
+    ));
+
+    Ok(gmp_api)
+}
+
+impl<T: GmpApiTrait + Send + Sync + 'static, U: GMPTaskAudit + Send + Sync + 'static, V: GMPAudit + Send + Sync + 'static>
+    GmpApiTrait for GmpApiDbAuditDecorator<T, U, V>
+{
```

Comment:
```suggestion
impl<T, U, V> GmpApiTrait for GmpApiDbAuditDecorator<T, U, V>
where
    T: GmpApiTrait  + Send + Sync + 'static,
    U: GMPTaskAudit + Send + Sync + 'static,
    V: GMPAudit     + Send + Sync + 'static,
{
```

Could also be applied in a few other places for slightly better readability.

Reply:
@themicp Love it! How about this? https://github.com/commonprefix/axelar-relayer/pull/51/commits/aa9cae2f7b4e11a763966eda1fb5d43e56a60b9d

If we end up with more of those "shorthands" for traits, we can have a new file, outside of utils.rs for it